File: examples/Cache.java
// ----------
// Cache.java
// ----------

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

final class Cache {
    public static void main (String[] args) {
        System.out.println("Cache.java");

        {
        int i = 2;
        int j = 2;
        assert i == j;
        ++i;
        assert i != j;
        ++j;
        assert i == j;
        }

        {
        Integer x = new Integer(2);
        Integer y = new Integer(2);
        assert x            != y;
        assert x.intValue() == y.intValue();
        ++x;
        assert x            != y;
        assert x.intValue() != y.intValue();
        ++y;
        assert x            == y;
        assert x.intValue() == y.intValue();
        }

        {
        Integer x = 2;
        Integer y = 2;
        assert x == y;
        ++x;
        assert x != y;
        }

        {
        Integer x = 128; // cache: [-128, 127]
        Integer y = 128;
        assert x != y;
        --x;
        --y;
        assert x == y;
        }

        {
        Long x = 128L;
        Long y = 128L;
        assert x != y;
        --x;
        --y;
        assert x == y;
        }

        {
        int  i = 2;
//      Long x = 2;        // doesn't compile
        Long x = (long) i;
        }

        {
        long    j = 2;
//      Integer x = j;       // doesn't compile
        Integer x = (int) j;
        }

        {
        int[]     a = {2, 3, 4};
//      Integer[] b = a;         // doesn't compile
        Integer[] b = {2, 3, 4};
        }

        {
        String s = "abc";
        String t = "abc";
        assert s == t;
        }

        {
        String s = "abc";
        String t = "ab" + "c";
        assert s == t;
        }

        {
        String s = "abc";
        String u = "ab";
        String v = "c";
        String t = u + v;
        assert s != t;
        assert s.equals(t);
        }

        System.out.println("Done.");}}

File: examples/Cache.py
#!/usr/bin/env python

# --------
# Cache.py
# --------

print "Cache.py"

x = 2
y = 2
assert x is y
x += 1
assert x != y
y += 1
assert x is y

x = 2
y = 2 + 0
assert x is y

x = 257
y = 257
assert x is y

x = 257           # cache: [-6, 256]
y = 257 + 0
assert x is not y
assert x ==     y
x -= 1
assert x is not y
assert x !=     y
y -= 1
assert x is y
assert x == y

x = 2L
y = 2L
assert x is y

x = 2L
y = 2L + 0
assert x is not y
assert x ==     y

x = 2.34
y = 2.34
assert x is y

x = 2.34
y = 2.34 + 0
assert x is not y
assert x ==     y

s = "abc"
t = "abc"
assert s is t

s = "abc"
t = "ab" + "c"
assert s is t

s = "abc"
u = "ab"
v = "c"
t = u + v
assert s is not t
assert s ==     t

a = []
b = []
assert a is not b
assert a ==     b

a = ()
b = ()
assert a is b

print "Done."
